### ISSUE TITLE
Replay attack: solutions

### ISSUE THREAD

**Issue by arvicco** [ethereumclassic/README#3]
Some people expressed concerns over cross-chain [replay attack](https://www.reddit.com/r/ethereum/comments/4t2jfq/security_the_oldtonewchain_replay_attack_and_the/).

However, if you plan to use both altchains (ETH and ETC), it is quite possible to avoid replay attacks:
- First of all, it is advisable to use two separate sets of private keys for operations on each network: one set for Ethereum, another for Ethereum Classic. Don't reuse the same key on the other network, even though they have the same format. 
- Exchanges need to cleanly separate their ETH and ETC wallets by following [this procedure](https://np.reddit.com/r/ethereum/comments/4t30ew/poloniex_announces_their_plans_regarding_upcoming/d5edz7r).
- The easiest options for users to split legacy ETH funds are the following:
  - Run them through Shapeshift splitter: https://split.shapeshift.io/ (0.25% fee)
  - Deposit legacy ETH to Poloniex, funds will be split automatically
  - Mix legacy funds with (already separated) ETH/ETC as described [here](https://np.reddit.com/r/ethereum/comments/4t30ew/poloniex_announces_their_plans_regarding_upcoming/d5egr9g)
  - Split legacy funds using smart contracts:
    - Fund splitting contract with [instruction how to use it](https://www.reddit.com/r/EthereumClassic/comments/4trcgn/seperate_classic_and_new_ether/d5kmo9h),
    -  [EF-recommended split contract](https://www.reddit.com/r/ethereum/comments/4t9mzg/quick_hf_safety_tips_for_users/), 
    -  [Another contract by independent developer](https://gist.github.com/tjade273/acb0d443592287d4f2fc44955e126854).

**Comment by trapp** [comment]
The initial split is not enough. You have to keep the balance of your separate addresses at 0 on the respective other chain. As soon as it receives a single replayable value transfer all your transfers become replayable up to the total sum of the balance the address has.

One way to permanently be safe would be to only send funds to the separated addresses through a smart contract that routes the funds depending on the chain it resides on:

Write up of the issue and this solution:
https://medium.com/@timonrapp/how-to-deal-with-the-ethereum-replay-attack-3fd44074a6d8#.1jk4dj7g3

This works for exchanges but obviously doesn't work for every network participant. A normal user receives money from other users and he can't be sure they only send non-replayable transactions.

**Comment by whatisgravity** [comment]
I have a working proof of concept of this attack that can be tested on private networks if anyone needs a copy to help decide how to best defend against this.

**Comment by ethernomad** [comment]
I think after the hard fork block it is essential that we switch to a new starting nonce.

**Comment by whatisgravity** [comment]
There are contracts available that will check which side of the fork you are on and act as a gateway. This should prevent any issues. Once you use this to migrate your ETH to a newly created address all the threats related to replay attacks are gone.

**Comment by whatisgravity** [comment]
I would not recommend using the HFConditionalTransfer contract, it is based on a variable in the WhiteHatDAO that could change without announcement. It is a dangerous to use this contract. Seems like a hacky solution.

#How To

**Description**
The *RelaySafeSplit* Kraken contract is much better, it uses another contract as an oracle, the oracle was deployed and checked the range between a defined time period then locked the result so it can not be tampered with or changed.

The result to a publicly accessible variable and can not be altered. (Assuming they do not change the blockchain)

It allows you to solve the entirely resolve the issue in a single contract too.

**Instructions**
I recomend using the RelaySafeSplit() 

You can deploy the code yourself or just use contract found at 0x341e790174e3a4d35b65fdc067b6b5634a61caea it is not hard coded and can be used by anyone

1. Create 2 new accounts, 1 for the destination of your tokens on each side of the fork. It does not matter which client you use because the accounts/keys are valid on chain. 

2. Obtain the JSON ABI from http://etherscan.io/address/0x341e790174e3a4d35b65fdc067b6b5634a61caea#code

3. Load Ethereum Wallet, Select "Contracts", Scrolldown and Select "Watch Contract"

4. Put in the name, the address and paste in the JSON ABI 

5. Click the contract

6. Scroll down to "Write to contract", click the "Select Function" dropdown

7. Execute the function "split"

8. Paste in each of the 2 new addresses, specify the balance you want to split and send it. 

9. Now your coins are safely split and can be used from now on without any further need to worry about the issue.

**Comment by trapp** [comment]
@whatisgravity, please update the referenced address to the updated version of the contract: 0xaa1a6e3e6ef20068f7f8d8c835d2d22fd5116444

The version you linked was meant for internal use only and should only be used with accounts as targets. When it's used with contracts as destination there could edge cases be triggered that trap the value in the contract.

Please recommend the updated version which works with both accounts and contract destinations: http://etherscan.io/address/0xaa1a6e3e6ef20068f7f8d8c835d2d22fd5116444

**Comment by whatisgravity** [comment]
Thank you for alerting me. I will do that now.

**Comment by whatisgravity** [comment]
I overlooked that, I clearly  need to sleep, looking for it now that is very obvious. 

I see now though, since it uses send it would run out of gas on the transfer to a cotnract, fall out of the if else if statement, then return false leaving the money in the contract. 

That would be very unfortunate. Thankfully no one reported any issues yet, I will take care of it if they do. Thanks for catching that.

I'm curious, why did you not include a suicide as a back up?  You could have recorded the last address that entered and used it as the output of the suicide as a failsafe.

**Comment by sulkair** [comment]
I'm confused.  I placed all my ETHER into brain wallet addresses which I exclusively own the private keys to.  This was long before the hard fork.  How can I separate them into ETC and ETH?  If I put them on Poloinex or Kracken will they automatically be recognized as PRE-FORK, and I'll get both ETC and ETH balances?  I would prefer a method not using exchanges.  Thanks you!

**Comment by whatisgravity** [comment]
Unfortunately I do not know how Kraken is handling this. From experiences relayed to me Poloneix is registering transfers from addresses with both tokens as two separate deposits effectively splitting any balance for you. @trapp may be able to provide more details for us about Kraken's system and if it works in the same way.

Alternatively you can use the contract provided by Kraken. I explained above how to use this contract, if there is any confusion, let me know so I can assist you and update the guide accordingly. If you do not want to risk putting your tokens on an exchange I believe the contract supplied by Kraken was well made and is easy to understand.

**Comment by sulkair** [comment]
So I just transferred 1 ETH from the Ethereum-Wallet to Poloniex ETH.  ...Note, I haven't touched my Ethereum-Wallet for over 4 months, so the address has to be pre-fork.  The 1 ETH has been posted to my Poloniex account, but it has not been split into an equal amount of ETC on the exchange.  Yet - anyway.

**Comment by benjyz** [comment]
@arvicco if some exchanges are committed to honour ETC and some are not, they have massive counterparty risk. there is high negative, accumulating incentive not to honour the ETC fork. the idea that things can be split cleanly mistaken. the way to resolve inconsistencies will be to void ETC **or** ETH balances. so the marketvalue of ETC/ETH can be seen as a probability of which chain will be honoured. According to this reasoning chance are 90% that ETH balances will be honoured and ETC voided. an ETC token is nothing else than an ETH token honouring the DAOattack balance. an ETH token is nothing else than an ETC token and not honouring the DAOattack balance. and post fork, these two states are inconsistent.

**Comment by benjyz** [comment]
@arvicco (let me rephrase): 

An ETC token is nothing else than an ETH token honouring the DAOattack balance. an ETH token is nothing else than an ETC token and not honouring the DAOattack balance. and post fork, these two states are inconsistent. Replay attack is related to double spend. Its only that same players have created the fiction that ETC balances can have value.

> If the users withdraws (already separated) ETH/ETC funds from the exchanges post-fork, s/he can cleanly "split" his other legacy funds by mixing funds as described here.

No, there is the double spending problem. There is no clean split here.

**Comment by arvicco** [comment]
@benjyz Obviously, major exchanges such as Bitfinex, Kraken and Poloniex (which collectively represent 70%+ of ETH trading) thought about risks clearly, and decided that this does not prevent them from offering ETC trading option. All these exchanges properly split their legacy funds and suffered no losses. It is only the exchanges that heeded the EF advice to "just ignore the legacy chain" that have suffered losses, and some of them are now trying to push such losses to their customers. 

I don't think such strategy would be effective, ultimately - at least not for the exchanges that are situated in normal jurisdictions and can be sued.

**Comment by arvicco** [comment]
"An ETC token is nothing else than an ETH token honouring the DAOattack balance" - this is simply not correct. Since the blockchains diverged, their tokens diverged as well. Mixing up double spending attack and replay problem is disingenuous.

**Comment by picojoule** [comment]
is this method still valid today? Shapeshift.io and Jaxx.io have discontinued their split feature for unknown reasons and Poloniex doesn't seem to be splitting anymore. I tried to follow the steps in the following article which afaik is the recommended method:

https://steemit.com/ethereum/@pauls/ethereum-fork-step-by-step-guide-to-safely-splitting-your-eth-etc

but have run into issues where the ETH arrives but the ETC does not. 

Really hoping someone can guide me as to what the recommended way is to do this.

**Comment by sulkair** [comment]
I'm not sure if this is the best way - but it worked for me.  I made two brand new unrelated addresses. I used: https://classicetherwallet.com/ for the ETC, and I used: https://www.myetherwallet.com for the ETH.  Then I used these same websites to send the ETH to my new ETH address, and the corresponding ETC website to send to my new ETC address.  I did it quickly.  And I checked my balances through the whole thing with https://etherscan.io for ETH and http://gastracker.io/ for ETC.   NOW IT IS ENTIRELY POSSIBLE THIS WAS DANGEROUS ( I hope someone else will chime in as to the risk ) But it did work for me and I moved quickly once I exposed the private key on the ETH (which I always moved first) to hurry up and move the ETC.  I did this on about 10 different addresses.  Worked great.

**Comment by arvicco** [comment]
Right now both ETH and ETC implemented replay protection for transactions. Pretty much all the client software supports it by default now, including https://classicetherwallet.com and https://www.myetherwallet.com. So, unless you specifically want to send a transaction in replayable format (or using some extremely old client SW), your transactions are not replayable.

**Comment by picojoule** [comment]
Even for ETH that existed before the ETC fork happened? I was hoping I
would be able to sell the ETC part and then buy more ETH like others have.
Are you saying this isnt possible anymore? Was there a window that i missed?

**Comment by sulkair** [comment]
you can absolutely still do that

**Comment by picojoule** [comment]
I got it done via myclassicetherwallet! Thanks!

**Comment by Marloe1** [comment]
Hi, 
He're propably not a very smart question. I use a Jaxx wallet and my ethereum was send to  0xAA1A6e3e6EF20068f7F8d8C835d2D22fd5116444 ( in november 2016) . 

Now it's gone from my wallet. How do I get my ether back?
